### PR TITLE
The War Within

### PR DESCRIPTION
**Pull request by mrbuds**
# Description

Pull Request for TWW

This will let the bot create packages in comments

# TODO

- Models Thumbnail on WeakAurasDisplayButton + on New Aura menu doesn't show the big sword guy
- AuraBar's Model subregion "Clipped by Progress" option doesn't work on TWW (fine on DF)
- pvp talent: use talentId instead of indexes because blizzard tend to shuffle them between patches https://github.com/WeakAuras/WeakAuras2/issues/4557#issuecomment-1663922194

### CODE REVIEW

**Comment by Stanzilla** [comment]
✨✨ Here's [an AI-assisted sketch of how you might approach this issue](https://copilot-workspace.githubnext.com/WeakAuras/WeakAuras2/pull/5071?shareId=7a2c90e0-9dcc-48ae-a051-1879146dd3c9) saved by @Stanzilla using Copilot Workspace v0.17

<details open><summary><b>Topic</b></summary>
<br/>

Analyze this PR and check for mistakes

</details>

<details open><summary><b>Before</b></summary>
<br/>

- The PR introduces changes to multiple files, including `.luacheckrc`, `WeakAuras/BuffTrigger2.lua`, `WeakAuras/Compatibility.lua`, `WeakAuras/GenericTrigger.lua`, `WeakAuras/Init.lua`, `WeakAuras/Prototypes.lua`, and others, affecting functionality related to WeakAuras addon for World of Warcraft.
- The PR updates interface version numbers in `.toc` files to `110000`.
- It adds a new file `WeakAuras/Compatibility.lua` to handle functions that work differently across WoW versions.
- Changes in `WeakAuras/BuffTrigger2.lua` and other Lua files replace direct `GetSpellInfo` calls with `WeakAuras.GetSpellInfo` to utilize the newly introduced compatibility layer.
- The PR removes specific Wrath of the Lich King Classic checks, consolidating version checks under broader categories like `IsCataOrRetail`.
- Adjustments are made to talent-related functionality, reflecting changes in the game's API and data structures in different WoW versions.
- The PR aims to ensure compatibility and functionality of the WeakAuras addon across multiple versions of World of Warcraft, including Classic Era, Cataclysm Classic, and Retail.

</details>
<details open><summary><b>After</b></summary>
<br/>

* The PR correctly updates interface version numbers in `.toc` files to `110000`.
* The new file `WeakAuras/Compatibility.lua` properly handles functions across different WoW versions.
* Direct `GetSpellInfo` calls are replaced with `WeakAuras.GetSpellInfo` to utilize the compatibility layer.
* The consolidation of version checks under broader categories like `IsCataOrRetail` is accurate.
* Adjustments to talent-related functionality reflect the game's API and data structures across WoW versions.


</details>

<details open><summary><b>Plan</b></summary>
<br/>

* `WeakAuras/Compatibility.lua` (CHANGE)
  - Add compatibility layer for `GetSpellInfo` across different WoW versions.
* `.luacheckrc` (CHANGE)
  - Add `C_Reputation` to globals section.
* `WeakAuras/BuffTrigger2.lua` (CHANGE)
  - Replace `GetSpellInfo` with `WeakAuras.GetSpellInfo` for compatibility across WoW versions.
  - Adjust talent-related functionality to reflect changes in the game's API and data structures.
* `WeakAuras/GenericTrigger.lua` (CHANGE)
  - Update function calls to use the new compatibility layer functions.
  - Remove specific Wrath of the Lich King Classic checks.
* `WeakAuras/Init.lua` (CHANGE)
  - Update interface version numbers in `.toc` files to `110000`.
  - Consolidate version checks under broader categories like `IsCataOrRetail`.
* `WeakAuras/Prototypes.lua` (CHANGE)
  - Adjust talent-related functionality to use the compatibility layer.
* `WeakAuras/Types.lua` (CHANGE)
  - Update powerTypes and other version-specific checks to use the compatibility layer.
* `WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasMiniTalent_Dragonflight.lua` (CHANGE)
  - Ensure compatibility with the new talent system structure.
* `WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasMultiLineEditBox.lua` (CHANGE)
  - Replace `GetSpellInfo` with `WeakAuras.GetSpellInfo` for drag-and-drop functionality.
* `WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasMultiLineEditBoxWithEnter.lua` (CHANGE)
  - Update to use `WeakAuras.GetSpellInfo` for drag-and-drop functionality.
* `WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasSpinBox.lua` (CHANGE)
  - Adjust UI offsets and widths for compatibility across WoW versions.
* `WeakAurasOptions/LoadOptions.lua` (CHANGE)
  - Replace `GetSpellInfo` with `WeakAuras.GetSpellInfo` in spell name correction function.
* `WeakAurasOptions/OptionsFrames/IconPicker.lua` (CHANGE)
  - Use `WeakAuras.GetSpellInfo` for fetching spell icons.
* `WeakAurasOptions/OptionsFrames/OptionsFrame.lua` (CHANGE)
  - Adjust UI element positions for compatibility across WoW versions.
* `WeakAurasTemplates/TriggerTemplatesData.lua` (CHANGE)
  - Update to use `WeakAuras.GetSpellInfo` for spell data.

</details>

<details open><summary><b>Sketch of implementation</b></summary>
<br/>

An implementation sketch has not been generated or was not included

<details open><summary><b>Details</b></summary>
<br/>

Code analyzed at e0fb057bae2018908eaae9090ba27ba26416d7f8

</details>

**Comment by mrbuds** [comment]
> Got this error when was switching between spell triggers, especially whenever I clicked `Charges Changed Event` or `Cooldown Ready Event`. Triggers worked as usual though

Thanks for the report, this should be fixed now

**Comment by ljosberinn** [comment]
- `GetItemSetInfo` in `Prototypes.lua` now needs to be `C_Item.GetItemSetInfo`
- It'd be nice of `Compatibility.lua` would include something for `GetSpellCooldown` (now `C_Spell.GetSpellCooldown`)

**Comment by mrbuds** [comment]
>     * It'd be nice of `Compatibility.lua` would include something for `GetSpellCooldown` (now `C_Spell.GetSpellCooldown` and also returns a table)

Your question raise a future problem, if people start using them, then it is weakauras addon burden to keep them working forever.
These helper functions were only made for internal usage and may be deleted without warning when all versions of the game handle them the same way.
I'd prefer move all functions of Compatibility.lua to private namespace before it is an issue.

File: WeakAuras/Prototypes.lua [hunk 1]
```diff
@@ -5012,8 +5085,7 @@ Private.event_prototypes = {
       }
     },
     iconFunc = function(trigger)
-      local _, _, icon = GetSpellInfo(trigger.spellName or 0);
-      return icon;
+      return Private.ExecEnv.GetSpellInfo(trigger.spellName or 0);
```
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:5088]: you want GetSpellIcon here

**Review by InfusOnWoW — CHANGES_REQUESTED** [review]

File: WeakAuras/Types_Retail.lua [hunk 2]
```diff
@@ -143,7 +143,7 @@ local backgroundAlias = {
 }
 
 local classIDToOffsets = {
-	[1] = { extraOffsetX = 30, extraOffsetY = 31, }, -- Warrior
+	[1] = { extraOffsetX = 60, extraOffsetY = 31, }, -- Warrior
```
[the remark this answers is not in the document]
Comment (mrbuds) [reply WeakAuras/Types_Retail.lua:146]: yes value is different in tww, i'll make it conditional

**Review by InfusOnWoW — APPROVED** [review]

**Comment by Peter-Day** [comment]
> Hi,
> 
> I noticed that running the `C_TooltipInfo.GetHyperlink()` function inside of WeakAuras (or in a Plater script) is throwing an error ("unable to find mixin (ColorMixin)") even though the function never changed for TWW. The function runs with no errors if executed via the chat with `/run` or `/dump` or if directly put into an addon's code. I'm sorry if this isn't an issue with WeakAuras but I don't know why the same command runs fine in one place but fails in another :|
> 
> So basically running this from chat works: `/dump C_TooltipInfo.GetHyperlink ("unit:" .. UnitGUID("player"))`
> 
> But putting this in a WeakAura throws the error "unable to find mixin (ColorMixin)": `local tooltipInfo = C_TooltipInfo.GetHyperlink ("unit:" .. UnitGUID("player"))`
> 
> PS: It fails with any GUID, I just use the player's here for convenience.
> 
> Here's the full error from a WeakAura that has just that one line of code:
> 
> ```
> 1x Lua error in aura 'Hyperlink WA': init
> WeakAuras Version: 5.13.1-15-g3a05188
> unable to find mixin (ColorMixin)
> [string "=[C]"]: in function `GetHyperlink'
> [string "return function() C_TooltipInfo.GetHyperlink ("unit:" .. UnitGUID("player"))"]:1: in function <[string "return function() C_TooltipInfo.GetHyperlin..."]:1>
> [string "=[C]"]: in function `xpcall'
> [string "@WeakAuras/AuraEnvironment.lua"]:374: in function `ActivateAuraEnvironment'
> [string "@WeakAuras/GenericTrigger.lua"]:1063: in function `ScanWithFakeEvent'
> [string "@WeakAuras/GenericTrigger.lua"]:1413: in function `LoadDisplays'
> [string "@WeakAuras/WeakAuras.lua"]:2030: in function `LoadDisplays'
> [string "@WeakAuras/WeakAuras.lua"]:1812: in function <WeakAuras/WeakAuras.lua:1649>
> [string "@WeakAuras/WeakAuras.lua"]:2007: in function `Resume'
> [string "@WeakAurasOptions/OptionsFrames/OptionsFrame.lua"]:170: in function <...dOns/WeakAurasOptions/OptionsFrames/OptionsFrame.lua:150>
> [string "=[C]"]: in function `Hide'
> [string "@Blizzard_UIParentPanelManager/Mainline/UIParentPanelManager.lua"]:1046: in function <...ParentPanelManager/Mainline/UIParentPanelManager.lua:1041>
> [string "@DevTool/Libs/AceConfig-3.0-3/AceConfigDialog-3.0-86/AceConfigDialog-3.0.lua"]:1846: in function <...nfig-3.0/AceConfigDialog-3.0/AceConfigDialog-3.0.lua:1845>
> [string "=[C]"]: in function `securecall'
> [string "@Blizzard_UIParentPanelManager/Mainline/UIParentPanelManager.lua"]:1087: in function `CloseWindows'
> [string "@Blizzard_UIParentPanelManager/Mainline/UIParentPanelManager.lua"]:1101: in function <...ParentPanelManager/Mainline/UIParentPanelManager.lua:1099>
> [string "=[C]"]: in function `securecall'
> [string "@Blizzard_UIParent/Mainline/UIParent.lua"]:2841: in function `ToggleGameMenu'
> [string "TOGGLEGAMEMENU"]:1: in function <[string "TOGGLEGAMEMENU"]:1>
> 
> Locals:
> (*temporary) = "unable to find mixin (ColorMixin)"
> ```

I'm getting the same error trying to use the function "C_ChallengeMode.GetKeystoneLevelRarityColor" from within any weakaura.

**Pull request merged**